### PR TITLE
[viu:ott] Add support to download entire series with --yes-playlist

### PR DESCRIPTION
**Pull request by lkho**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Dear maintainers:
**Fix:**
Passing `--yes-playlist` is identical to not passing any `--xxx-playlist` option since there is a default False value.
**Improvement:**
The video info got from `viu:ott` has all the video ids of the entire series. So maybe we can support downloading of all the episodes with the `--yes-playlist` flag.

### CODE REVIEW

File: youtube_dl/options.py [hunk 1]
```diff
@@ -326,11 +326,11 @@ def _comma_separated_values_options_callback(option, opt_str, value, parser):
         ))
     selection.add_option(
         '--no-playlist',
-        action='store_true', dest='noplaylist', default=False,
```
Comment (dstftw) [inline youtube_dl/options.py]: No.

File: youtube_dl/options.py [hunk 2]
```diff
@@ -326,11 +326,11 @@ def _comma_separated_values_options_callback(option, opt_str, value, parser):
...
+        action='store_true', dest='noplaylist',
         help='Download only the video, if the URL refers to a video and a playlist.')
     selection.add_option(
         '--yes-playlist',
-        action='store_false', dest='noplaylist', default=False,
```
Comment (dstftw) [inline youtube_dl/options.py]: No.

File: youtube_dl/extractor/viu.py [hunk 3]
```diff
@@ -168,7 +168,7 @@ def _real_extract(self, url):
 
 class ViuOTTIE(InfoExtractor):
     IE_NAME = 'viu:ott'
-    _VALID_URL = r'https?://(?:www\.)?viu\.com/ott/(?P<country_code>[a-z]{2})/[a-z]{2}-[a-z]{2}/vod/(?P<id>\d+)'
+    _VALID_URL = r'(?P<scheme>viuott|https?)://(?:www\.)?viu\.com/ott/(?P<country_code>[a-z]{2})/(?P<lang_code>[a-z]{2}-[a-z]{2})/vod/(?P<id>\d+)'
```
Comment (dstftw) [inline youtube_dl/extractor/viu.py]: No such scheme.

File: youtube_dl/extractor/viu.py [hunk 4]
```diff
@@ -210,6 +210,32 @@ def _real_extract(self, url):
...
+                    if not item_id:
+                        continue
+                    item_id = compat_str(item_id)
+                    entries.append(self.url_result(
+                        'viuott://www.viu.com/ott/%s/%s/vod/%s/' % (country_code, lang_code, item_id),
+                        'ViuOTT',
+                        item_id,
+                        entry.get('synopsis', '').strip()))
+
+                self.to_screen('Downloading all episodes in series %s because of --yes-playlist' % video_data.get('series_id'))
```
Comment (dstftw) [inline youtube_dl/extractor/viu.py]: No. Must be the opposite since `--yes-playlist` is default.

File: youtube_dl/extractor/viu.py [hunk 5]
```diff
@@ -210,6 +210,32 @@ def _real_extract(self, url):
...
+            product = series.get('product')
+            if product:
+                entries = []
+                for entry in sorted(product, key=lambda x: int_or_none(x.get('number', 0))):
+                    item_id = entry.get('product_id')
+                    if not item_id:
+                        continue
+                    item_id = compat_str(item_id)
+                    entries.append(self.url_result(
+                        'viuott://www.viu.com/ott/%s/%s/vod/%s/' % (country_code, lang_code, item_id),
```
Comment (dstftw) [inline youtube_dl/extractor/viu.py]: No. Use `smuggle_url` if you need to pass some additional attribute.

**Comment by dstftw** [comment]
>Passing `--yes-playlist` is identical to not passing any `--xxx-playlist` option since there is a default False value.

Yes and you've just broken it with this PR. Now the default is `None`.

Also add a test.

**Comment by lkho** [comment]
@dstftw 
> Yes and you've just broken it with this PR. Now the default is `None` that means not specified.

Does this mean the `--yes-playlist` flag is actually not necessary and is only for documentation purpose in which this flag should be identical to not passing? Regarding this case, would you suggest changing the default to `None` (as I didn't find any code that will break) or there is another way to switch the preferred behaviour?

**Comment by dstftw** [comment]
>Does this mean the `--yes-playlist` flag is actually not necessary and is only for documentation purpose in which this flag should be identical to not passing?

No, it does not mean that.

>Regarding this case, would you suggest changing the default to `None` (as I didn't find any code that will break) or there is another way to switch the preferred behaviour?

No, don't touch the default behavior.

**Comment by lkho** [comment]
i think if the `--yes-playlist` flag is useful then the default value of `noplaylist` should be `None`?

**Comment by dstftw** [comment]
No, it does not mean that. The default value is `False` and it won't change.

**Comment by lkho** [comment]
maybe --yes-playlist should add a variable `yesplaylist`? or else the flag makes no difference.

**Comment by dstftw** [comment]
**No.** It makes perfect difference, e.g. for overriding `--no-playlist` defined in more generic wide configuration.

**Comment by lkho** [comment]
ok i got it. how about the original purpose of downloading a playlist if both video and playlist is available, such as the entire series of the current episode?

**Comment by dstftw** [comment]
How what? I've already pointed out what should be fixed.

**Comment by lkho** [comment]
I am just thinking the url is actually referring to one episode, while there is no specific url for "all episodes". So the user will always need to specify `--no-playlist` even the url does not look like a playlist (unlike youtube the url has a separate `?list=` id besides the `?v=`).

Maybe adding a new extractor with a different name and specify it in the options?

Anyway, I will update the logic according to your suggestion if there is really no other way to improve this default flag.

**Comment by lkho** [comment]
any updates on this?

**Comment by lkho** [comment]
Hi @dstftw , could you kindly review this if you have time? Just wanna see if there are still problems.

**Comment by zackmark29** [comment]
Is there any update for this issue?
It would be nice if you could add support to download the entire series

**Comment by lkho** [comment]
The maintainer said this is intended behavior regarding the playlist flags, and will not change it. So if you want to download viu:ott as a list, either patch it yourself (e.g. fork my branch), or change the viu:ott extractor to ALWAYS return a playlist. Thus you can use `--no-playlist` every time you want a single episode, OR, omit the flag to have a full playlist.

If I didn't misunderstand @dstftw's opinion.

**Comment by lkho** [comment]
anyway i am still wondering how to handle if the url represents a single video, but it also expose a list. How can the user choose between?

**Comment by zackmark29** [comment]
> The maintainer said this is intended behavior regarding the playlist flags, and will not change it. So if you want to download viu:ott as a list, either patch it yourself (e.g. fork my branch), or change the viu:ott extractor to ALWAYS return a playlist. Thus you can use `--no-playlist` every time you want a single episode, OR, omit the flag to have a full playlist.
> 
> If I didn't misunderstand @dstftw's opinion.



> The maintainer said this is intended behavior regarding the playlist flags, and will not change it. So if you want to download viu:ott as a list, either patch it yourself (e.g. fork my branch), or change the viu:ott extractor to ALWAYS return a playlist. Thus you can use `--no-playlist` every time you want a single episode, OR, omit the flag to have a full playlist.
> 
> If I didn't misunderstand @dstftw's opinion.

ahm could you give me an example? I don't really understand :)
Here's the link  that I want to download all series at once
https://www.viu.com/ott/ph/en-us/vod/121181/The-Producers#tab2

**Comment by lkho** [comment]
**Short answer**
you cannot

**long answer**
the current `viu:ott` extractor does NOT return playlist result, although it has the playlist info already. To make it work, you must either MODIFY the code of either youtube-dl itself, or `viu:ott` extractor to return a playlist result instead.

The latest compiled binaries released by youtube-dl officially will NOT have the things you want, and you will have to download Python on your computer and run it locally, if you are to modify the code and run the customized program.

**Comment by lkho** [comment]
>  `https://www.viu.com/ott/id/id/all/video-chinese-drama-movies-back_to_1942-1164195806`

currently this extractor can only extract all the episodes of ONE entire series, such as `https://www.viu.com/ott/hk/zh-hk/vod/259033/`, and does not recognize any other pages including `https://www.viu.com/ott/hk/zh-hk/category/4/` or whatever other listings e.g. `https://www.viu.com/ott/hk/zh-hk/all`.

so it is intentionally not matched.

**Comment by lkho** [comment]
Actually this PR has been here for 2.5 years already. The original suggestion is to give an option to download all the items of a series, since the series itself has NO standalone listing page, even until now 2020. Correct me if I am wrong.

**Comment by zackmark29** [comment]
> > `https://www.viu.com/ott/id/id/all/video-chinese-drama-movies-back_to_1942-1164195806`
> 
> currently this extractor can only extract all the episodes of ONE entire series, such as `https://www.viu.com/ott/hk/zh-hk/vod/259033/`, and does not recognize any other pages including `https://www.viu.com/ott/hk/zh-hk/category/4/` or whatever other listings e.g. `https://www.viu.com/ott/hk/zh-hk/all`.
> 
> so it is intentionally not matched.

In my personal Viu tool, I can get the playlists. I don't know why youtube-dl doesn't have this feature
[https://i.imgur.com/WINmha1.jpg](url)

**Comment by lkho** [comment]
> I don't know why youtube-dl doesn't have this feature

because there is yet no one submitted the code regarding the said feature here.
of course you can submit another PR if you have figured out how to do that, or a feature request to see if someone could try.

**Comment by zackmark29** [comment]
> > I don't know why youtube-dl doesn't have this feature
> 
> because there is yet no one submitted the code regarding the said feature here.
> of course you can submit another PR if you have figured out how to do that, or a feature request to see if someone could try.

I only code on C#. I don't know yet python.
Wait, what are you talking about is the "Getting the entire playlists of the series?" 
Or could you give me an example link and I'll check

**Comment by lkho** [comment]
what this PR  does:

given `https://www.viu.com/ott/ph/en-us/vod/286677/Flower-of-Evil`
return all the urls of episode 1 to 10, instead of just 286677 (ep1)

**Comment by zackmark29** [comment]
> what this PR does:
> 
> given `https://www.viu.com/ott/ph/en-us/vod/286677/Flower-of-Evil`
> return all the urls of episode 1 to 10, instead of just 286677 (ep1)
> 
> just like the dialog of your screenshot

Yeah, only 1 episode needed and automatically getting all the series links

**Comment by lkho** [comment]
take an example from my region (HK), which has 11 episodes currently:

```
C:\Users\lkho>youtube-dl https://www.viu.com/ott/hk/zh-hk/vod/254942/blablablablabla --get-url
https://stream-hk.viu.com/s/u8rqE9DiegA6KiRcxpjhoA/1599076637/UD/27645ce30c28a42e4d656dc1a1451242/27645ce30c28a42e4d656dc1a1451242_Layer4_vod.m3u8
https://stream-hk.viu.com/s/JoeNJ5AXUfBtG8nP2Eadtw/1599076638/UD/b42d3d204a6458b2d4e70b0bda7c7d4b/b42d3d204a6458b2d4e70b0bda7c7d4b_Layer4_vod.m3u8
https://stream-hk.viu.com/s/jJ0dlvmDrs_h02NqZopg5A/1599076639/UD/3ce7b96cffc5cb95d686dd652e16a042/3ce7b96cffc5cb95d686dd652e16a042_Layer4_vod.m3u8
https://stream-hk.viu.com/s/jJ0dlvmDrs_h02NqZopg5A/1599076639/UD/d66b9aa053940aa45b2ccace060b0cae/d66b9aa053940aa45b2ccace060b0cae_Layer4_vod.m3u8
https://stream-hk.viu.com/s/sZhuxJQvB_v16EmN41k2Qg/1599076640/UD/5c81d999e5a8296fa568be330b3eb2b5/5c81d999e5a8296fa568be330b3eb2b5_Layer4_vod.m3u8
https://stream-hk.viu.com/s/M3B8m9Hrp5gooBcDS3cUuQ/1599076641/UD/25ee0cd9689dd720c0a1e269fa3b8dc7/25ee0cd9689dd720c0a1e269fa3b8dc7_Layer4_vod.m3u8
https://stream-hk.viu.com/s/4oBCflvL8wkCgVqO0VWNvA/1599076642/UD/2b012655cd8dd4014361b5c2c17df423/2b012655cd8dd4014361b5c2c17df423_Layer4_vod.m3u8
https://stream-hk.viu.com/s/u1HHsbBET34B6U7dVldXvw/1599076646/UD/d429363047e56010a9d0a540d2ad7a40/d429363047e56010a9d0a540d2ad7a40_Layer4_vod.m3u8
https://stream-hk.viu.com/s/wGIkQW67MvYceHUJPzJDGw/1599076647/UD/835d423e7750b8c7cac2ea9fb00c1021/835d423e7750b8c7cac2ea9fb00c1021_Layer4_vod.m3u8
https://stream-hk.viu.com/s/8IYWSuRdTKPJM_UhPfeSrQ/1599076650/UD/00e28ceaebba81671af1751f914cc6fa/00e28ceaebba81671af1751f914cc6fa_Layer4_vod.m3u8
https://stream-hk.viu.com/s/XdHrt7GDfaizf2ssBkTO-w/1599076651/UD/bceadb5fdcda2688c4706977a075ea38/bceadb5fdcda2688c4706977a075ea38_Layer4_vod.m3u8

C:\Users\lkho>
```

so there are 11 .m3u8 urls already, corresponding to the 11 episodes

**Pull request closed without merging**